web/client/epics/queryparams.js

```diff
@@ -58,18 +59,22 @@ const paramActions = {
     },
     center: (parameters, state) => {
         const map = mapSelector(state);
+        const { heading, pitch, roll = 0 } = parameters;
+        const validViewerOptions = [heading, pitch].map(val => typeof(val) !== 'undefined');
         const validCenter = parameters && !isEmpty(parameters.center) && parameters.center.split(',').map(val => !isEmpty(val) && toNumber(val));
         const center = validCenter && validCenter.indexOf(false) === -1 && getCenter(validCenter);
         const zoom = toNumber(parameters.zoom);
         const bbox =  getBbox(center, zoom);
         const mapSize = map && map.size;
         const projection = map && map.projection;
-        const viewerOptions = map.viewerOptions;
+        const viewerOptions = validViewerOptions && validViewerOptions.indexOf(false) === -1 ? { heading, pitch, roll } : map.viewerOptions;
         const isValid = center && isObject(center) && inRange(center.y, -90, 91) && inRange(center.x, -180, 181) && inRange(zoom, 1, 36);
-
-        if (isValid) {
+        if (isValid && !viewerOptions) {
             return [changeMapView(center, zoom, bbox, mapSize, null, projection, viewerOptions)];
         }
+        if (isValid && viewerOptions) {
+            return [changeMapType('cesium'), changeMapView(center, zoom, bbox, mapSize, null, projection, viewerOptions)];
```

Comment:
Here we are assuming that if viewerOptions exists we switch to cesium so it seems that viewerOptions is always related to 3D, is it always true?

I'm asking this because in the previous lines where we define viewerOptions we are taking the viewerOptions from the map config

Reply:
> Here we are assuming that if viewerOptions exists we switch to cesium so it seems that viewerOptions is always related to 3D, is it always true?


@allyoucanmap , apparently yes, if you have a look at the `updateMapInfoState` method (triggered upon every map move) for openlayers the `viewerOptions` is explicitly set to undefined each time the map moves (line 476 in `MapStore2\web\client\components\map\openlayers\Map.jsx`) 

For leaflet I see coming from the `props' (line 443 in `MapStore2\web\client\components\map\leaflet\Map.jsx`) 
 but I see no place in the code where this property is set (with a reducer for example)